File: gauntlet/java/Alias.java
public class Alias {
    static void smooth(double[] dst, double[] src) {
        for (int i = 1; i < src.length - 1; i++)
            dst[i] = (src[i-1] + src[i] + src[i+1]) / 3;
    }
    static void smoothNoAlias(double[] dst, double[] src) {
        double a = src[0], b = src[1];
        for (int i = 1; i < src.length - 1; i++) {
            double c = src[i+1];
            dst[i] = (a + b + c) / 3;
            a = b; b = c;
        }
    }
    static void bench(String name, Runnable fn, int iters) {
        for (int i = 0; i < iters * 3; i++) fn.run();
        double[] ts = new double[7];
        for (int r = 0; r < 7; r++) {
            long t0 = System.nanoTime();
            for (int i = 0; i < iters; i++) fn.run();
            ts[r] = (System.nanoTime() - t0) / (double) iters;
        }
        java.util.Arrays.sort(ts);
        System.out.printf("%-30s %10.0f ns/op%n", name, ts[3]);
    }
    public static void main(String[] a) {
        int N = 1 << 16;
        double[] src = Gauntlet.makeVec(N, 7);
        double[] dst = new double[N];
        bench("smooth disjoint", () -> smooth(dst, src), 300);
        bench("smoothNoAlias disjoint", () -> smoothNoAlias(dst, src), 300);
        bench("smooth IN-PLACE (aliased)", () -> smooth(src, src), 300);
        bench("smoothFresh (allocates)", () -> smooth(new double[N], src), 300);
    }
}

File: gauntlet/js/alias.mjs
import { makeVec } from "./gauntlet.mjs";

function smooth(dst, src) {
  for (let i = 1; i < src.length - 1; i++)
    dst[i] = (src[i - 1] + src[i] + src[i + 1]) / 3;
}
function smoothNoAlias(dst, src) {
  let a = src[0], b = src[1];
  for (let i = 1; i < src.length - 1; i++) {
    const c = src[i + 1];
    dst[i] = (a + b + c) / 3;
    a = b; b = c;
  }
}
function bench(name, fn, iters) {
  for (let i = 0; i < iters; i++) fn();
  const ts = [];
  for (let r = 0; r < 7; r++) {
    const t0 = process.hrtime.bigint();
    for (let i = 0; i < iters; i++) fn();
    ts.push(Number(process.hrtime.bigint() - t0) / iters);
  }
  ts.sort((a, b) => a - b);
  console.log(`${name.padEnd(30)} ${ts[3].toFixed(0).padStart(10)} ns/op`);
}
const N = 1 << 16;
const src = makeVec(N, 7);
const dst = new Float64Array(N);
bench("smooth disjoint", () => smooth(dst, src), 200);
bench("smoothNoAlias disjoint", () => smoothNoAlias(dst, src), 200);
bench("smooth IN-PLACE (aliased)", () => smooth(src, src), 200);
bench("smoothFresh (allocates)", () => { const d = new Float64Array(N); smooth(d, src); }, 200);
